diesel/src/r2d2.rs

```diff
@@ -0,0 +1,216 @@
+//! Connection pooling via r2d2
+
+extern crate r2d2;
+
+pub use self::r2d2::*;
+
+use std::convert::Into;
+use std::fmt;
+use std::marker::PhantomData;
+
+use backend::UsesAnsiSavepointSyntax;
+use deserialize::QueryableByName;
+use prelude::*;
+use connection::{AnsiTransactionManager, SimpleConnection};
+use query_builder::{AsQuery, QueryFragment, QueryId};
+use sql_types::HasSqlType;
+
+/// An r2d2 connection manager for use with Diesel.
+///
+/// See the [r2d2 documentation] for usage examples.
+///
+/// [r2d2 documentation]: ../../r2d2
+#[derive(Debug, Clone)]
+pub struct ConnectionManager<T> {
+    database_url: String,
+    _marker: PhantomData<T>,
+}
+
+unsafe impl<T: Send + 'static> Sync for ConnectionManager<T> {}
+
+impl<T> ConnectionManager<T> {
+    /// Returns a new connection manager,
+    /// which establishes connections to the given database URL.
+    pub fn new<S: Into<String>>(database_url: S) -> Self {
+        ConnectionManager {
+            database_url: database_url.into(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// The error used when managing connections with `r2d2`.
+#[derive(Debug)]
+pub enum Error {
+    /// An error occurred establishing the connection
+    ConnectionError(ConnectionError),
+
+    /// An error occurred pinging the database
+    QueryError(::result::Error),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::ConnectionError(ref e) => e.fmt(f),
+            Error::QueryError(ref e) => e.fmt(f),
+        }
+    }
+}
+
+impl ::std::error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::ConnectionError(ref e) => e.description(),
+            Error::QueryError(ref e) => e.description(),
+        }
+    }
+}
+
+impl<T> ManageConnection for ConnectionManager<T>
+where
+    T: Connection + Send + 'static,
+{
+    type Connection = T;
+    type Error = Error;
+
+    fn connect(&self) -> Result<T, Error> {
+        T::establish(&self.database_url).map_err(Error::ConnectionError)
+    }
+
+    fn is_valid(&self, conn: &mut T) -> Result<(), Error> {
+        conn.execute("SELECT 1")
+            .map(|_| ())
+            .map_err(Error::QueryError)
+    }
+
+    fn has_broken(&self, _conn: &mut T) -> bool {
+        false
+    }
+}
+
+impl<T> SimpleConnection for PooledConnection<ConnectionManager<T>>
+where
+    T: Connection + Send + 'static,
+{
+    fn batch_execute(&self, query: &str) -> QueryResult<()> {
+        (&**self).batch_execute(query)
+    }
+}
+
+impl<C> Connection for PooledConnection<ConnectionManager<C>>
+where
+    C: Connection<TransactionManager = AnsiTransactionManager> + Send + 'static,
+    C::Backend: UsesAnsiSavepointSyntax,
```

Comment:
Why this constrain is required?
This code should work for any kind of backend as far as I see

Reply:
This is explained in both the commit message and PR description.

Reply:
>  If we wanted to support anything that
> implemented `diesel::Connection`, we would need to add the constraint
> `C::TransactionManager: TransactionManager<Self>`. However,
> `TransactionManager` has the bound that it's parameter be `Connection`,
> so we just recurse infinitely. Maybe on day the language will be smarter
> about this. Until then, we just have to do this weird hack.